lib/web_ui/lib/src/engine/canvas_pool.dart

```diff
@@ -0,0 +1,694 @@
+      _secondaryAttributesDirty = true;
+    }
+  }
+
+  set strokeCap(ui.StrokeCap strokeCap) {
+    strokeCap ??= ui.StrokeCap.butt;
+    if (strokeCap != _currentStrokeCap) {
+      _currentStrokeCap = strokeCap;
+      context.lineCap = _stringForStrokeCap(strokeCap);
+      _secondaryAttributesDirty = true;
+    }
+  }
+
+  set lineWidth(double lineWidth) {
+    if (lineWidth != _currentLineWidth) {
+      _currentLineWidth = lineWidth;
+      context.lineWidth = lineWidth;
+      _secondaryAttributesDirty = true;
+    }
+  }
+
+  set strokeJoin(ui.StrokeJoin strokeJoin) {
+    strokeJoin ??= ui.StrokeJoin.miter;
+    if (strokeJoin != _currentStrokeJoin) {
+      _currentStrokeJoin = strokeJoin;
+      context.lineJoin = _stringForStrokeJoin(strokeJoin);
+      _secondaryAttributesDirty = true;
+    }
+  }
+
+  set fillStyle(Object colorOrGradient) {
+    if (!identical(colorOrGradient, _currentFillStyle)) {
+      _currentFillStyle = colorOrGradient;
+      context.fillStyle = colorOrGradient;
+      if (assertionsEnabled) {
+        _currentFillStyle = context.fillStyle;
+      }
+    }
+  }
+
+  set strokeStyle(Object colorOrGradient) {
+    if (!identical(colorOrGradient, _currentStrokeStyle)) {
+      _currentStrokeStyle = colorOrGradient;
+      context.strokeStyle = colorOrGradient;
+    }
+  }
+
+  set filter(String filter) {
+    if (_currentFilter != filter) {
+      _currentFilter = filter;
+      context.filter = filter;
+      _secondaryAttributesDirty = true;
+    }
+  }
+
+  void paint(ui.PaintingStyle style) {
+    if (style == ui.PaintingStyle.stroke) {
+      context.stroke();
+    } else {
+      context.fill();
+    }
+  }
+
+  void reset() {
+    context.fillStyle = '';
+    _currentFillStyle = context.fillStyle;
+    context.strokeStyle = '';
+    _currentStrokeStyle = context.strokeStyle;
+    context.filter = 'none';
+    _currentFilter = 'none';
+    context.globalCompositeOperation = 'source-over';
+    _currentBlendMode = ui.BlendMode.srcOver;
+    context.lineWidth = 1.0;
+    _currentLineWidth = 1.0;
+    context.lineCap = 'butt';
+    _currentStrokeCap = ui.StrokeCap.butt;
+    context.lineJoin = 'miter';
+    _currentStrokeJoin = ui.StrokeJoin.miter;
+  }
+}
+
+/// Provides save stack tracking functionality to implementations of
+/// [EngineCanvas].
+class _SaveStackTracking {
+  static final Vector3 _unitZ = Vector3(0.0, 0.0, 1.0);
+
+  final List<_SaveStackEntry> _saveStack = <_SaveStackEntry>[];
+
+  /// The stack that maintains clipping operations used when text is painted
+  /// onto bitmap canvas but is composited as separate element.
+  List<_SaveClipEntry> _clipStack;
+
+  /// Returns whether there are active clipping regions on the canvas.
+  bool get isClipped => _clipStack != null;
+
+  /// Empties the save stack and the element stack, and resets the transform
+  /// and clip parameters.
+  ///
+  /// Classes that override this method must call `super.clear()`.
```

Comment:
We can enforce `super.clear()` by annotating it with `@mustCallSuper` (also applies to `save`, `restore`, and other methods in this class)

Reply:
Done.